Shane/make stateful

shanemachine92 opened the pull request:
I've updated the webcrawler to be stateful. Instead of storing and tracking urls in memory, it now writes to a SQLite3 databse.

Commit:
prevent duplicates

Commit:
write tests for document collection

Commit:
test url collector

Commit:
test document class

Commit:
fix crawler tests

Commit:
puts crawler on loop, rescue HTTP :errors

Commit:
create db write methods

Commit:
purge url collection

Commit:
rescue errors, crawl relative urls

Commit:
save documents to database

Commit:
delete obsolete files, methods, variables

Commit:
avoid duplicates to db

Commit:
set notification when done crawling

Commit:
fix error

JesseBuesking commented on Rakefile:
```diff
@@ -0,0 +1,4 @@
+task :run_crawler do
+  puts "rake task running"
```
Probably don't need this `puts`, or at least you could have something more descriptive here.

JesseBuesking commented on Rakefile:
```diff
@@ -0,0 +1,4 @@
+task :run_crawler do
+  puts "rake task running"
+  `ruby /Users/shane/work/Morrigan/script.rb`
```
Can you make this path in some way relative. As it is, this will only work on your computer.

shanemachine92 replied:
Probably going to delete this rake task all together, as it's not adding a ton of value

JesseBuesking commented on crawler.rb:
```diff
@@ -1,39 +1,63 @@
-	end
+  attr_accessor :url, :href, :db
+  def initialize(db)
+    @database = db 
```
I *think* db is a database connection in this case. If so, we typically call it `conn` or `connection`.

shanemachine92 replied:
To clarify, do you mean that the method should be def connection instead of def initialize?

JesseBuesking replied:
The variable `db` and the instance variable `@database` could be called `conn` and `@conn` instead.

shanemachine92 replied:
Got it! Thanks!

JesseBuesking commented on crawler.rb:
```diff
@@ -1,39 +1,63 @@
+  end
+
+  def done_crawling?
+    @database.execute( "SELECT count(*) FROM URLS_to_crawl WHERE state = 'uncrawled'" )[0][0] == 0
```
No whitespace around between parentheses.

JesseBuesking replied:
Also, after this PR if it's not already on your roadmap, you should look into using rails models with migrations. I think this week you're working on learning SQL so this makes sense as-is.

JesseBuesking commented on crawler.rb:
```diff
@@ -1,39 +1,63 @@
+
+  def crawl_next_url
+    begin
+      url = @database.execute( "SELECT url FROM URLS_to_crawl WHERE state = 'uncrawled' LIMIT 1" )[0][0]
```
I'd toss an `order by` on this.

shanemachine92 replied:
Really? Order by what?

JesseBuesking commented on crawler.rb:
```diff
@@ -1,39 +1,63 @@
+      content = UrlFetcher.fetch(url)
+      document = Document.new(url, content)
+      write_documents_to_db(document)
+      document.domain_hrefs.each do |href|
```
You can one-line this. You may *also* want to put this in a transaction.

cbnoonan replied:
I second a lot of jesse's  comments, but especially this one. :)

shanemachine92 replied:
Working on wrapping it in a transaction! Getting a "cannot start a transaction within a transaction" error (thanks error handling!), but will figure that out  and push when it's resolved.

JesseBuesking commented on crawler.rb:
```diff
@@ -1,39 +1,63 @@
+  end
+
+  def update_state(url)
+    @database.execute( "UPDATE URLS_to_crawl SET state = 'crawled' WHERE url = ?", [url] )
```
No ws in the rest of these parentheses.

I ❤️  that you're doing parameterized queries here.

JesseBuesking commented on crawler.rb:
```diff
@@ -1,39 +1,63 @@
+  def run
+    if done_crawling?
+      puts "Crawling complete!"
+      %x(/usr/local/bin/terminal-notifier -message "Total Documents: #{@database.execute("SELECT COUNT(*) FROM documents")}" -title "Crawling Complete!")
```
Lol, you're using your notifier! I dig it.

JesseBuesking commented on script.rb:
```diff
@@ -1,20 +1,28 @@
+db.execute <<-SQL
+  CREATE TABLE IF NOT EXISTS documents (
+    url varchar(200) PRIMARY KEY,
+    content varchar(65535)
```
You could possibly change this from `varchar(65535)` to `TEXT`. Up to you tho.

cbnoonan replied:
I agree with TEXT here, it's more commonly used. If you wanted to check the differences between varchar and TEXT, I support that study!

shanemachine92 replied:
Done! Didn't know about TEXT :)

JesseBuesking commented on script.rb:
```diff
@@ -1,20 +1,28 @@
 
+db.execute <<-SQL
+  CREATE TABLE IF NOT EXISTS documents (
+    url varchar(200) PRIMARY KEY,
```
Can `url` be longer? What does the standard specify, if there is one?

shanemachine92 replied:
It can be in theory. Changed this to type TEXT

JesseBuesking commented on spec/crawler_spec.rb:
```diff
@@ -0,0 +1,22 @@
+require_relative '../script'
+
+RSpec.describe Crawler do
+	subject(:crawler) { Crawler.new(document_collection, url_collection)}
```
Your ws looks off in this spec.

JesseBuesking commented on spec/crawler_spec.rb:
```diff
@@ -0,0 +1,22 @@
+	# let (:url_fetcher){UrlFetcher.new}
+	# let (:content) {url_fetcher.fetch(:url)}
+	# let(:document) {Document.new(:url, :content)}
+	let(:document_collection) {DocumentCollection.new}
```
In the case of `{}` for lambda blocks, you *do* want to have a space in between.

JesseBuesking commented on spec/crawler_spec.rb:
```diff
@@ -0,0 +1,22 @@
+	# let (:content) {url_fetcher.fetch(:url)}
+	# let(:document) {Document.new(:url, :content)}
+	let(:document_collection) {DocumentCollection.new}
+	let(:url_collection) {BreadthFirstUrlCollection.new}
```
I think you removed `BreadthFirstUrlCollection`, so this shouldn't be necessary.

shanemachine92 replied:
Wow, this file is a mess. I removed all of these classes, so this test doesn't make sense anymore. Will remove.

JesseBuesking commented on spec/document_spec.rb:
```diff
@@ -0,0 +1,12 @@
+require_relative '../script'
+
```
ws off in here too.

JesseBuesking commented on crawler.rb:
```diff
@@ -1,39 +1,63 @@
+                VALUES (?, 'uncrawled')", ['http://dragonage.wikia.com' + url])
+  end  
+
+  def write_documents_to_db(document)
```
This method is grabbing the next url, fetching the content, and inserting the content into the db. It might make sense to rename this method to something more appropriate.

JesseBuesking commented on crawler.rb:
```diff
@@ -1,39 +1,63 @@
+  end  
+
+  def write_documents_to_db(document)
+    url = @database.execute( "SELECT url FROM URLS_to_crawl WHERE state = 'uncrawled' LIMIT 1" )[0][0]
```
You should probably have an `order by` here as well. I like being as explicit as possible in my queries, even if the defaults are sane. Just me tho.

JesseBuesking commented on crawler.rb:
```diff
@@ -1,39 +1,63 @@
+
+  def write_documents_to_db(document)
+    url = @database.execute( "SELECT url FROM URLS_to_crawl WHERE state = 'uncrawled' LIMIT 1" )[0][0]
+    content = UrlFetcher.fetch(url)
```
Should you check the `documents` table first to see if the url is there? If not, why do you need to `INSERT OR IGNORE`? Based on that I assume that the url *might* be in `documents`.

shanemachine92 replied:
You're right, there's already an ```INSERT OR IGNORE``` at the ```write_urls_to_database``` level that protects against duplicates. It's unnecessary here, so will remove.

JesseBuesking commented on document.rb:
```diff
@@ -2,27 +2,26 @@
-		@url = url
-		@content = content
-	end
+  def initialize(url, content)
```
Woohoo, better ws!

JesseBuesking reviewed (changes requested):
I also think that you should have a unique index on your `URLS_to_crawl` table. Your code seems to make this assumption, and having a database constraint to guarantee this makes it clear within the database itself.

cbnoonan commented on script.rb:
```diff
@@ -1,20 +1,28 @@
-require_relative './depth_first_url_collection'
 
-BASE_URL = 'http://dragonage.wikia.com'
+db = SQLite3::Database.new 'webcrawler.db'
```
you might want to include a README of how this script is called somewhere

cbnoonan reviewed (changes requested):
@JesseBuesking  did a great job capturing most things.

shanemachine92 committed:
apply first round feedback

shanemachine92 committed:
apply first round feedback

shanemachine92 committed:
create transaction

JesseBuesking commented on crawler.rb:
```diff
@@ -7,55 +7,55 @@
-        write_urls_to_db(href)
-      end
+      get_url_and_write_document(document)
+      @connection.transaction
```
I think this would be better as a block call. What happens if you iterator below raises an exception mid-transaction?

shanemachine92 committed:
jesse feedback

shanemachine92 commented:
@JesseBuesking I'd missed your unique index suggestion. Added one and updated the transaction block. Let me know if that's what you had in mind or if I should make further changes.

JesseBuesking commented on crawler.rb:
```diff
@@ -1,39 +1,64 @@
+
+  def crawl_next_url
+    begin
+      @connection.transaction
```
Why not just do

```ruby
@connection.transaction do
  '... things to be put in a transaction here ...'
end
```

If your code fails to throw a `StandardError` or `SQLite3::Exception`, your transaction will not rollback. Doing it within the block ensures that the `rollback` will occur. You may be asking yourself "When will this ever happen?". TBH I dunno, but I'd rather not have to think about it and would prefer to use their safer block-supporting syntax to take care of it for me :smile:

If you REAAAALLLLY wanna know, then crack open the source for rails and see how they implement it. There's no better way to understand it than to peel back the magic and see what they're doing. 99% of the time you'll be like "Oh, that's it?", but that's ok. Sometimes you'll be like "Oh god why???". That's never fun.

`StandardError` is basically a catch-all too, so there's really no reason to have both `StandardError` and `SQLite3::Exception` catching and running the same block.

shanemachine92 commented:
Haha I thought about doing it that way @JesseBuesking but I couldn't really discern why it was better.

And thanks for the note on the errors. I wasn't sure if SQLite errors were included in standard. Although I could have googled it : P

I'm reading Ruby Under a Microscope now and it's so interesting!! I'll dig into the rails source code soon.

Thank for you the continued feedback!

JesseBuesking commented:
Oh snap, I hope you enjoy `Ruby Under a Microscope` as much as I did 😄 

No worries. My bad if there are delays in my feedback btw. In the upcoming weeks it'll unfortunately only get worse but I'll continue to try my best to follow up.

shanemachine92 committed:
add transaction block

JesseBuesking reviewed (approved):
Awesome! Congrats on your first rebase too 👍  🍾
